# 05_231211_RePRSTrials

### 主目的

Tonga集団のサンプルを用いて倹約遺伝子仮説を検討する。

1. 過去のPRSを推定する：今回はその前段階として現在ベースでPRSを推定する。
2. ポリジェニックセレクションの検討：これから（まず手法をもう一度チェック！）

### 実行にあたって

Tonga集団のデータを用いてPRSを計算したい。が、いま手元に彼らの身長や体重に関するデータがないことがわかった。

とりあえずその前段階として遺伝統計学夏の学校2023の実習を再現してみる。

### 遺伝統計学夏の学校2023【QC以降の手順確認】

まずはデータファイルをダウンロード・解凍する。

In [12]:
mkdir -p ~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
cd ~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
ls

In [13]:
wget http://www.sg.med.osaka-u.ac.jp/files/StatGenSummerSchool2023_GenomeDataAnalysis5.zip
ls

--2023-12-18 16:22:57--  http://www.sg.med.osaka-u.ac.jp/files/StatGenSummerSchool2023_GenomeDataAnalysis5.zip
Resolving www.sg.med.osaka-u.ac.jp (www.sg.med.osaka-u.ac.jp)... 133.1.138.171
Connecting to www.sg.med.osaka-u.ac.jp (www.sg.med.osaka-u.ac.jp)|133.1.138.171|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 185582439 (177M) [application/zip]
Saving to: ‘StatGenSummerSchool2023_GenomeDataAnalysis5.zip’

StatGenSummerSchool 100%[===================>] 176.98M  21.7MB/s    in 6.7s    

2023-12-18 16:23:04 (26.3 MB/s) - ‘StatGenSummerSchool2023_GenomeDataAnalysis5.zip’ saved [185582439/185582439]

StatGenSummerSchool2023_GenomeDataAnalysis5.zip


In [14]:
unzip StatGenSummerSchool2023_GenomeDataAnalysis5.zip

Archive:  StatGenSummerSchool2023_GenomeDataAnalysis5.zip
   creating: StatGenSummerSchool2023_GenomeDataAnalysis5/
   creating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/
   creating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/


  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/1KG_EUR_QC.bed  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/1KG_EUR_QC.bim  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/1KG_EUR_QC.fam  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/phenotype_RA.txt  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/RA_GWASmeta_European_PRSise.txt.gz  
   creating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output/
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output/1KG_EUR-RAGWAS-PRSboxplot.png  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output/1KG_EUR-RAGWAS.best  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output/1KG_EUR-RAGWAS.best.txt  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output/1KG_EUR-RAGWAS.log  
  inflating: StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output/1KG_EUR-RAGWAS.

In [15]:
ls

StatGenSummerSchool2023_GenomeDataAnalysis5
StatGenSummerSchool2023_GenomeDataAnalysis5.zip


In [16]:
ls StatGenSummerSchool2023_GenomeDataAnalysis5

PRSice  StatGenSummerSchool2023_GenomeDataAnalysis5.pdf


In [17]:
ls StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice

Input                      PRsice_Command.txt  PRSice_mac.R
Output                     PRSice_linux        PRSice.R
PRsice_Command_cygwin.txt  PRSice_linux.R      PRSice_win64.exe
PRsice_Command_mac.txt     PRSice_mac


In [18]:
ls StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input
ls StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Output

1KG_EUR_QC.bed  1KG_EUR_QC.fam    RA_GWASmeta_European_PRSise.txt.gz
1KG_EUR_QC.bim  phenotype_RA.txt


1KG_EUR-RAGWAS_BARPLOT_2019-08-25.png        1KG_EUR-RAGWAS.log
1KG_EUR-RAGWAS_BARPLOT_2021-07-14.png        1KG_EUR-RAGWAS-PRSboxplot.png
1KG_EUR-RAGWAS.best                          1KG_EUR-RAGWAS.prsice
1KG_EUR-RAGWAS.best.txt                      1KG_EUR-RAGWAS.summary
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2019-08-25.png  1KG_EUR_Sample.txt
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2021-07-14.png  PlotPRSice.R


#### 入力ファイルの確認

In [19]:
ls StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input

1KG_EUR_QC.bed  1KG_EUR_QC.fam    RA_GWASmeta_European_PRSise.txt.gz
1KG_EUR_QC.bim  phenotype_RA.txt


In [20]:
zcat StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/RA_GWASmeta_European_PRSise.txt.gz | head -n 10

SNP	CHR	BP	A1	A2	P	OR
chr1:751343	1	751343	A	T	0.01	0.85
chr1:751756	1	751756	T	C	0.01	1.17
rs3094315	1	752566	A	G	0.0093	1.14
rs3131972	1	752721	A	G	0.009	0.88
rs3131971	1	752894	T	C	0.01	0.87
chr1:753405	1	753405	A	C	0.01	1.17
chr1:753425	1	753425	T	C	0.0097	0.87
rs2073814	1	753474	G	C	0.009	1.14
rs2073813	1	753541	A	G	0.0095	0.85


In [21]:
cat StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice/Input/phenotype_RA.txt | head -n 10

HG00096	HG00096	2
HG00097	HG00097	1
HG00099	HG00099	1
HG00100	HG00100	1
HG00101	HG00101	2
HG00102	HG00102	2
HG00103	HG00103	1
HG00104	HG00104	1
HG00106	HG00106	1
HG00108	HG00108	1


#### PRS解析の実行

In [22]:
cd ~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials/StatGenSummerSchool2023_GenomeDataAnalysis5/PRSice
mkdir -p Output_2
touch Output_2/.gitkeep
ls

Input     PRsice_Command_cygwin.txt  PRSice_linux    PRSice_mac.R
Output    PRsice_Command_mac.txt     PRSice_linux.R  PRSice.R
Output_2  PRsice_Command.txt         PRSice_mac      PRSice_win64.exe


In [23]:
Rscript PRSice_linux.R \
--prsice PRSice_linux \
--dir . \
--out ./Output_2/1KG_EUR-RAGWAS \
--base ./Input/RA_GWASmeta_European_PRSise.txt.gz \
--target ./Input/1KG_EUR_QC \
--pheno ./Input/phenotype_RA.txt \
--thread 1 \
--stat OR \
--binary-target T

Trying to install data.table in ./lib
Trying to install optparse in ./lib
PRSice 2.3.3 (2020-08-05) 
https://github.com/choishingwan/PRSice
(C) 2016-2020 Shing Wan (Sam) Choi and Paul F. O'Reilly
GNU General Public License v3
If you use PRSice in any published work, please cite:
Choi SW, O'Reilly PF.
PRSice-2: Polygenic Risk Score Software for Biobank-Scale Data.
GigaScience 8, no. 7 (July 1, 2019)
2023-12-18 16:23:34
./PRSice_linux \
    --a1 A1 \
    --a2 A2 \
    --bar-levels 0.001,0.05,0.1,0.2,0.3,0.4,0.5,1 \
    --base ./Input/RA_GWASmeta_European_PRSise.txt.gz \
    --binary-target T \
    --bp BP \
    --chr CHR \
    --clump-kb 250kb \
    --clump-p 1.000000 \
    --clump-r2 0.100000 \
    --interval 5e-05 \
    --lower 5e-08 \
    --num-auto 22 \
    --or  \
    --out ./Output_2/1KG_EUR-RAGWAS \
    --pheno ./Input/phenotype_RA.txt \
    --pvalue P \
    --seed 2986104432 \
    --snp SNP \
    --stat OR \
    --target ./Input/1KG_EUR_QC \
    --thread 1 \
    --upper 0.5

Init

#### Output結果の確認

In [24]:
ls Output_2

1KG_EUR-RAGWAS_BARPLOT_2023-12-18.png        1KG_EUR-RAGWAS.log
1KG_EUR-RAGWAS.best                          1KG_EUR-RAGWAS.prsice
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-18.png  1KG_EUR-RAGWAS.summary


In [25]:
head Output_2/1KG_EUR-RAGWAS.prsice

Pheno	Set	Threshold	R2	P	Coefficient	Standard.Error	Num_SNP
-	Base	5e-08	0.0122466	0.0629842	12.6059	6.77991	257
-	Base	5.005e-05	0.0126982	0.0583659	22.8194	12.055	479
-	Base	0.00010005	0.01382	0.0484141	27.829	14.0998	561
-	Base	0.00015005	0.0139415	0.0474483	30.2877	15.2792	611
-	Base	0.00020005	0.0143518	0.0443525	33.0976	16.4604	658
-	Base	0.00025005	0.0146751	0.0420519	35.1525	17.2909	691
-	Base	0.00030005	0.015445	0.0370909	37.9226	18.1904	729
-	Base	0.00035005	0.0160981	0.0333494	41.2045	19.3644	778
-	Base	0.00040005	0.0162487	0.0325501	43.4408	20.3223	817


In [26]:
head Output_2/1KG_EUR-RAGWAS.summary

Phenotype	Set	Threshold	PRS.R2	Full.R2	Null.R2	Prevalence	Coefficient	Standard.Error	P	Num_SNP
-	Base	0.46	0.0512432	0.0512432	0	-	5081.42	1356.18	0.000179058	101385


In [27]:
#画像表示準備
mkdir -p ~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials/pic
cp Output_2/*.png ~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials/pic/

【1KG_EUR-RAGWAS_BARPLOT_2023-12-11.png（左）】【1KG_EUR-RAGWAS_BARPLOT_2023-12-11.png（右）】

<img src="pic/1KG_EUR-RAGWAS_BARPLOT_2023-12-11.png" width="49%">
<img src="pic/1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-11.png" width="49%">

#### PRS解析結果の検討

In [28]:
head Output_2/1KG_EUR-RAGWAS.best

FID IID In_Regression PRS
HG00096 HG00096 Yes -0.000627027476
HG00097 HG00097 Yes -0.00058856461
HG00099 HG00099 Yes -0.000713382756
HG00100 HG00100 Yes -0.000571751231
HG00101 HG00101 Yes -0.000530233701
HG00102 HG00102 Yes -0.000583765514
HG00103 HG00103 Yes -0.000573780098
HG00104 HG00104 Yes -0.000506927325
HG00106 HG00106 Yes -0.000537212649


In [29]:
#ファイルの調整
awk '$1!="FID" && $3="Yes" {print $1"\t"$2"\t"$4}' Output_2/1KG_EUR-RAGWAS.best > Output_2/1KG_EUR-RAGWAS.best.txt

In [30]:
cat Output_2/1KG_EUR-RAGWAS.best.txt | head -n 10

HG00096	HG00096	-0.000627027476
HG00097	HG00097	-0.00058856461
HG00099	HG00099	-0.000713382756
HG00100	HG00100	-0.000571751231
HG00101	HG00101	-0.000530233701
HG00102	HG00102	-0.000583765514
HG00103	HG00103	-0.000573780098
HG00104	HG00104	-0.000506927325
HG00106	HG00106	-0.000537212649
HG00108	HG00108	-0.000588234975


In [31]:
ls *

PRsice_Command_cygwin.txt  PRSice_linux    PRSice_mac.R
PRsice_Command_mac.txt     PRSice_linux.R  PRSice.R
PRsice_Command.txt         PRSice_mac      PRSice_win64.exe

Input:
1KG_EUR_QC.bed  1KG_EUR_QC.fam    RA_GWASmeta_European_PRSise.txt.gz
1KG_EUR_QC.bim  phenotype_RA.txt

lib:
data.table  getopt  optparse

Output:
1KG_EUR-RAGWAS_BARPLOT_2019-08-25.png        1KG_EUR-RAGWAS.log
1KG_EUR-RAGWAS_BARPLOT_2021-07-14.png        1KG_EUR-RAGWAS-PRSboxplot.png
1KG_EUR-RAGWAS.best                          1KG_EUR-RAGWAS.prsice
1KG_EUR-RAGWAS.best.txt                      1KG_EUR-RAGWAS.summary
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2019-08-25.png  1KG_EUR_Sample.txt
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2021-07-14.png  PlotPRSice.R

Output_2:
1KG_EUR-RAGWAS_BARPLOT_2023-12-18.png        1KG_EUR-RAGWAS.log
1KG_EUR-RAGWAS.best                          1KG_EUR-RAGWAS.prsice
1KG_EUR-RAGWAS.best.txt                      1KG_EUR-RAGWAS.summary
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-18.png


In [32]:
#Rスクリプトの調整
cat Output/PlotPRSice.R -n

     1	#ディレクトリの移動
     2		rm(list = ls(all = TRUE))
     3		setwd("C:/SummerSchool/GenomeDataAnalysis5/PRSice/Output");
     4	
     5	
     6		PRS <- read.table("1KG_EUR-RAGWAS.best.txt", header=FALSE)[,3];
     7		Pop <- read.table("1KG_EUR_Sample.txt", header=FALSE)[,2];
     8	
     9		png(file="1KG_EUR-RAGWAS-PRSboxplot.png", width = 600, height = 600);
    10		par(mar = c(5, 5, 0.5, 0.5));
    11		boxplot(PRS~Pop, col=c(2,3,4,5,6), pch=16, xlab="1KG EUR sub-populations", ylab="Polygenic risk score for rheumatoid arthritis", cex.lab=2.0, cex.axis=2.0, cex.main=2.0);
    12		dev.off();
    13		
    14		anova(aov(PRS~Pop)) # P-value for Pop < 2.2x10^-16
    15		


In [33]:
cat Output/PlotPRSice.R | sed 1,5d

	PRS <- read.table("1KG_EUR-RAGWAS.best.txt", header=FALSE)[,3];
	Pop <- read.table("1KG_EUR_Sample.txt", header=FALSE)[,2];

	png(file="1KG_EUR-RAGWAS-PRSboxplot.png", width = 600, height = 600);
	par(mar = c(5, 5, 0.5, 0.5));
	boxplot(PRS~Pop, col=c(2,3,4,5,6), pch=16, xlab="1KG EUR sub-populations", ylab="Polygenic risk score for rheumatoid arthritis", cex.lab=2.0, cex.axis=2.0, cex.main=2.0);
	dev.off();
	
	anova(aov(PRS~Pop)) # P-value for Pop < 2.2x10^-16
	


In [34]:
echo 'setwd("Output_2")' > Output_2/PlotPRSice.R
cat Output/PlotPRSice.R | sed 1,5d >> Output_2/PlotPRSice.R
ls Output_2

1KG_EUR-RAGWAS_BARPLOT_2023-12-18.png        1KG_EUR-RAGWAS.log
1KG_EUR-RAGWAS.best                          1KG_EUR-RAGWAS.prsice
1KG_EUR-RAGWAS.best.txt                      1KG_EUR-RAGWAS.summary
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-18.png  PlotPRSice.R


In [35]:
cat Output_2/PlotPRSice.R

setwd("Output_2")
	PRS <- read.table("1KG_EUR-RAGWAS.best.txt", header=FALSE)[,3];
	Pop <- read.table("1KG_EUR_Sample.txt", header=FALSE)[,2];

	png(file="1KG_EUR-RAGWAS-PRSboxplot.png", width = 600, height = 600);
	par(mar = c(5, 5, 0.5, 0.5));
	boxplot(PRS~Pop, col=c(2,3,4,5,6), pch=16, xlab="1KG EUR sub-populations", ylab="Polygenic risk score for rheumatoid arthritis", cex.lab=2.0, cex.axis=2.0, cex.main=2.0);
	dev.off();
	
	anova(aov(PRS~Pop)) # P-value for Pop < 2.2x10^-16
	


In [36]:
#必要ファイルの準備
cp Output/1KG_EUR_Sample.txt Output_2/

In [37]:
#Rスクリプトの実行
Rscript Output_2/PlotPRSice.R

null device 
          1 
Analysis of Variance Table

Response: PRS
           Df     Sum Sq    Mean Sq F value    Pr(>F)    
Pop         4 6.4407e-07 1.6102e-07  33.237 < 2.2e-16 ***
Residuals 376 1.8215e-06 4.8440e-09                      
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


In [38]:
ls Output_2

1KG_EUR-RAGWAS_BARPLOT_2023-12-18.png        1KG_EUR-RAGWAS-PRSboxplot.png
1KG_EUR-RAGWAS.best                          1KG_EUR-RAGWAS.prsice
1KG_EUR-RAGWAS.best.txt                      1KG_EUR-RAGWAS.summary
1KG_EUR-RAGWAS_HIGH-RES_PLOT_2023-12-18.png  1KG_EUR_Sample.txt
1KG_EUR-RAGWAS.log                           PlotPRSice.R


In [39]:
#画像表示の準備
cp Output_2/1KG_EUR-RAGWAS-PRSboxplot.png ~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials/pic

<img src="pic/1KG_EUR-RAGWAS-PRSboxplot.png" width="49%">

### 東アジア集団GWAS（BaseData）のQC

まずややこしいのでコード格納ディレクトリとストレージディレクトリをそれぞれ変数にあてておく

In [40]:
gitdir=~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials
ls $gitdir

strdir=~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
ls $strdir

05_231211_RePRSTrials.ipynb  Documents  jupyter_venv5  pic  test.ipynb
StatGenSummerSchool2023_GenomeDataAnalysis5
StatGenSummerSchool2023_GenomeDataAnalysis5.zip


In [41]:
cd $strdir
mkdir -p Basedata
ls

Basedata
StatGenSummerSchool2023_GenomeDataAnalysis5
StatGenSummerSchool2023_GenomeDataAnalysis5.zip


#### データのダウンロード

論文はこれを使う。

https://www-nature-com.utokyo.idm.oclc.org/articles/s41588-017-0011-x

In [42]:
cd $strdir/Basedata
wget https://portals.broadinstitute.org/collaboration/giant/images/7/78/BMI_Eastern_Asian.fmt.gzip

--2023-12-18 16:24:36--  https://portals.broadinstitute.org/collaboration/giant/images/7/78/BMI_Eastern_Asian.fmt.gzip
Resolving portals.broadinstitute.org (portals.broadinstitute.org)... 69.173.68.142
Connecting to portals.broadinstitute.org (portals.broadinstitute.org)|69.173.68.142|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12268615 (12M)
Saving to: ‘BMI_Eastern_Asian.fmt.gzip’

BMI_Eastern_Asian.f 100%[===================>]  11.70M  5.43MB/s    in 2.2s    

2023-12-18 16:24:39 (5.43 MB/s) - ‘BMI_Eastern_Asian.fmt.gzip’ saved [12268615/12268615]



In [43]:
ls $strdir/Basedata

BMI_Eastern_Asian.fmt.gzip


In [44]:
head BMI_Eastern_Asian.fmt.gzip

CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100003875	G	A	rs148923672	-	A:0	NA	Inf	NA
10	100008694	C	G	rs200213822	-	G:0	1.6	1	0.1
10	100010909	C	T	rs141069497	T:0.0139	T:0.006614	-0.022	0.1	0.83
10	100011387	C	T	rs199733612	-	T:0,G:0	1.1	1	0.27
10	100011442	C	T	rs74153530	T:0	T:0,G:0	0.66	0.59	0.26
10	100013325	C	T	rs145270262	-	T:0	NA	Inf	NA
10	100013367	C	T	rs199557224	-	T:0	NA	Inf	NA
10	100013382	A	G	rs141995277	G:0	G:0	NA	Inf	NA
10	100013418	T	G	rs148260647	-	G:0	0.41	1	0.68


#### QCのチェックリスト
* Heritability Check
* Effect Allele
* File Transfer
* Genome Build
* Standard GWAS QC
* Mismatching SNPs
* Duplicate SNPs
* Ambiguous SNPs
* Sex Chromosomes
* Sample overlap
* Relatedness

##### Heritability Check

今回はスキップさせてくれ…　たぶん大丈夫だとは思うのだが。

（この前提が崩れるとそもそもこれが使えない。そういうものは公開されてないと思う。）

##### Effect Allele

MAFでフィルタリングするときに同時にチェックします。

##### File Transfer

ファイルの損傷チェック。とりあえずスキップ。

##### Genome Build：hg19/hg38が合っているかのチェック

In [45]:
head -n 2 BMI_Eastern_Asian.fmt.gzip

CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100003875	G	A	rs148923672	-	A:0	NA	Inf	NA


https://www.ncbi.nlm.nih.gov/snp/?term=rs148923672

上記を確認すると、

    10:98244118 (GRCh38)
    10:100003875 (GRCh37)
    
とあるので、このデータはGRCh37で間違いなさそう。

##### Standard GWAS QC：MAFやINFOが低いSNPは除去した方が良い！

In [46]:
cd $strdir/Basedata

In [47]:
head BMI_Eastern_Asian.fmt.gzip | \
sed -e s/A://g -e s/T://g -e s/G://g -e s/C://g | \
grep -v ,

CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100003875	G	A	rs148923672	-	0	NA	Inf	NA
10	100008694	C	G	rs200213822	-	0	1.6	1	0.1
10	100010909	C	T	rs141069497	0.0139	0.006614	-0.022	0.1	0.83
10	100013325	C	T	rs145270262	-	0	NA	Inf	NA
10	100013367	C	T	rs199557224	-	0	NA	Inf	NA
10	100013382	A	G	rs141995277	0	0	NA	Inf	NA
10	100013418	T	G	rs148260647	-	0	0.41	1	0.68


In [48]:
head BMI_Eastern_Asian.fmt.gzip | \
sed -e s/A://g -e s/T://g -e s/G://g -e s/C://g | \
grep -v , | \
awk 'NR==1 || ($6 > 0.01) {print}'

CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100010909	C	T	rs141069497	0.0139	0.006614	-0.022	0.1	0.83


In [49]:
head BMI_Eastern_Asian.fmt.gzip | \
grep -v , 

CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100003875	G	A	rs148923672	-	A:0	NA	Inf	NA
10	100008694	C	G	rs200213822	-	G:0	1.6	1	0.1
10	100010909	C	T	rs141069497	T:0.0139	T:0.006614	-0.022	0.1	0.83
10	100013325	C	T	rs145270262	-	T:0	NA	Inf	NA
10	100013367	C	T	rs199557224	-	T:0	NA	Inf	NA
10	100013382	A	G	rs141995277	G:0	G:0	NA	Inf	NA
10	100013418	T	G	rs148260647	-	G:0	0.41	1	0.68


In [50]:
head BMI_Eastern_Asian.fmt.gzip | \
grep -v , | \
sed -e s/:/"\t"/g -e s/-/-"\t"-/g

CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100003875	G	A	rs148923672	-	-	A	0	NA	Inf	NA
10	100008694	C	G	rs200213822	-	-	G	0	1.6	1	0.1
10	100010909	C	T	rs141069497	T	0.0139	T	0.006614	-	-0.022	0.1	0.83
10	100013325	C	T	rs145270262	-	-	T	0	NA	Inf	NA
10	100013367	C	T	rs199557224	-	-	T	0	NA	Inf	NA
10	100013382	A	G	rs141995277	G	0	G	0	NA	Inf	NA
10	100013418	T	G	rs148260647	-	-	G	0	0.41	1	0.68


In [51]:
echo -e 'CHR\tPOS\tREF\tALT\tSNPNAME\tEAS\tEAS_MAF\tExAC_EAS\tExAC_EAS_MAF\tbeta\tse\tPvalue'
head BMI_Eastern_Asian.fmt.gzip | \
grep -v , | \
sed -e s/:/"\t"/g -e s/-/-"\t"-/g

CHR	POS	REF	ALT	SNPNAME	EAS	EAS_MAF	ExAC_EAS	ExAC_EAS_MAF	beta	se	Pvalue


CHR	POS	REF	ALT	SNPNAME	EAS_MAF	ExAC_EAS_MAF	beta	se	Pvalue
10	100003875	G	A	rs148923672	-	-	A	0	NA	Inf	NA
10	100008694	C	G	rs200213822	-	-	G	0	1.6	1	0.1
10	100010909	C	T	rs141069497	T	0.0139	T	0.006614	-	-0.022	0.1	0.83
10	100013325	C	T	rs145270262	-	-	T	0	NA	Inf	NA
10	100013367	C	T	rs199557224	-	-	T	0	NA	Inf	NA
10	100013382	A	G	rs141995277	G	0	G	0	NA	Inf	NA
10	100013418	T	G	rs148260647	-	-	G	0	0.41	1	0.68


In [52]:
echo -e 'CHR\tPOS\tREF\tALT\tSNPNAME\tEAS\tEAS_MAF\tExAC_EAS\tExAC_EAS_MAF\tbeta\tse\tPvalue'
head BMI_Eastern_Asian.fmt.gzip | \
grep -v , | \
sed -e s/:/"\t"/g -e s/-/-"\t"-/g | \
sed 1d

CHR	POS	REF	ALT	SNPNAME	EAS	EAS_MAF	ExAC_EAS	ExAC_EAS_MAF	beta	se	Pvalue


10	100003875	G	A	rs148923672	-	-	A	0	NA	Inf	NA
10	100008694	C	G	rs200213822	-	-	G	0	1.6	1	0.1
10	100010909	C	T	rs141069497	T	0.0139	T	0.006614	-	-0.022	0.1	0.83
10	100013325	C	T	rs145270262	-	-	T	0	NA	Inf	NA
10	100013367	C	T	rs199557224	-	-	T	0	NA	Inf	NA
10	100013382	A	G	rs141995277	G	0	G	0	NA	Inf	NA
10	100013418	T	G	rs148260647	-	-	G	0	0.41	1	0.68


In [53]:
echo -e 'CHR\tPOS\tREF\tALT\tSNPNAME\tEAS\tEAS_MAF\tExAC_EAS\tExAC_EAS_MAF\tbeta\tse\tPvalue' > BMI_Eastern_Asian.newcol.txt
cat BMI_Eastern_Asian.fmt.gzip | \
grep -v , | \
sed -e s/:/"\t"/g -e s/-/-"\t"-/g | \
sed 1d >> BMI_Eastern_Asian.newcol.txt 

In [54]:
cat BMI_Eastern_Asian.newcol.txt | \
awk 'NR==1 || ($7 > 0.01) {print}' | wc

  31194  389695 1870772


In [55]:
cat BMI_Eastern_Asian.newcol.txt | \
awk 'NR==1 || ($7 > 0.01) && ($4 == $6) {print}' | wc

  31170  389376 1869585


In [56]:
cat BMI_Eastern_Asian.newcol.txt | \
awk 'NR==1 || ($7 > 0.01) && ($4 == $6) {print}' > BMI_Eastern_Asian.txt

In [57]:
head BMI_Eastern_Asian.txt

CHR	POS	REF	ALT	SNPNAME	EAS	EAS_MAF	ExAC_EAS	ExAC_EAS_MAF	beta	se	Pvalue
10	100010909	C	T	rs141069497	T	0.0139	T	0.006614	-	-0.022	0.1	0.83
10	100017453	T	G	rs1983864	G	0.6081	G	0.5682	-	-0.031	0.015	0.044
10	100017532	G	A	rs77858766	A	0.0169	A	0.01781	-	-0.016	0.058	0.79
10	100017553	G	T	rs11189525	T	0.0129	T	0.007404	-	-0.014	0.13	0.91
10	100020880	C	T	rs33995374	T	0.1062	T	0.1163	0.077	0.061	0.21
10	100044136	G	A	rs4919199	A	0.9544	-	-	0.0076	0.036	0.83
10	100144782	C	T	rs2296441	T	0.4425	T	0.42	0.012	0.015	0.42
10	100148004	T	C	rs2182169	C	0.5615	-	-	0.014	0.015	0.37
10	100148176	A	G	rs2147896	G	0.5833	G	0.5596	0.0052	0.015	0.73


In [58]:
ls -l BMI_Eastern_Asian.txt
wc BMI_Eastern_Asian.txt

-rw-rw-r-- 1 hmiwa hmiwa 1869585 12月 18 16:24 BMI_Eastern_Asian.txt


  31170  389376 1869585 BMI_Eastern_Asian.txt


##### Mismatching SNPs：ターゲットデータで対立遺伝子が一致しているかどうかのチェック。

ターゲットデータがないとチェックできないのでここはスキップ。

##### Duplicate SNPs

In [59]:
cat BMI_Eastern_Asian.txt |\
awk '{seen[$5]++; if(seen[$5]==1){print}}' > BMI_Eastern_Asian.nodup.txt

In [60]:
ls -l BMI_Eastern_Asian.nodup.txt
wc BMI_Eastern_Asian.nodup.txt

-rw-rw-r-- 1 hmiwa hmiwa 1869585 12月 18 16:24 BMI_Eastern_Asian.nodup.txt


  31170  389376 1869585 BMI_Eastern_Asian.nodup.txt


##### Ambiguous SNPs

In [61]:
head BMI_Eastern_Asian.nodup.txt

CHR	POS	REF	ALT	SNPNAME	EAS	EAS_MAF	ExAC_EAS	ExAC_EAS_MAF	beta	se	Pvalue
10	100010909	C	T	rs141069497	T	0.0139	T	0.006614	-	-0.022	0.1	0.83
10	100017453	T	G	rs1983864	G	0.6081	G	0.5682	-	-0.031	0.015	0.044
10	100017532	G	A	rs77858766	A	0.0169	A	0.01781	-	-0.016	0.058	0.79
10	100017553	G	T	rs11189525	T	0.0129	T	0.007404	-	-0.014	0.13	0.91
10	100020880	C	T	rs33995374	T	0.1062	T	0.1163	0.077	0.061	0.21
10	100044136	G	A	rs4919199	A	0.9544	-	-	0.0076	0.036	0.83
10	100144782	C	T	rs2296441	T	0.4425	T	0.42	0.012	0.015	0.42
10	100148004	T	C	rs2182169	C	0.5615	-	-	0.014	0.015	0.37
10	100148176	A	G	rs2147896	G	0.5833	G	0.5596	0.0052	0.015	0.73


In [62]:
cat BMI_Eastern_Asian.nodup.txt |\
awk '!( ($3=="A" && $4=="T") || \
        ($3=="T" && $4=="A") || \
        ($3=="G" && $4=="C") || \
        ($3=="C" && $4=="G")) {print}' > BMI_Eastern_Asian.QC.txt

In [63]:
ls -l BMI_Eastern_Asian.QC.txt
wc BMI_Eastern_Asian.QC.txt

-rw-rw-r-- 1 hmiwa hmiwa 1707833 12月 18 16:24 BMI_Eastern_Asian.QC.txt


  28519  356256 1707833 BMI_Eastern_Asian.QC.txt


##### Sex Chromosomes

やりようがないのでスルー

##### Sample overlap
オセアニア集団とは絶対重ならないのでOK

##### Relatedness

本来はやるべきだが今回は一旦スキップ

### Tonga集団サンプル（TargetData）のQC

そもそもデータはここにあります

In [2]:
#必要な変数チェック
gitdir=~/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials
echo $gitdir
ls $gitdir

strdir=~/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
echo $strdir
ls $strdir

#ストレージに移動
cd $strdir
echo $strdir
ls

/home/hmiwa/hmiwa/hmiwa-lab/d_20231023/05_231211_RePRSTrials
05_231211_RePRSTrials.ipynb  Documents  jupyter_venv5  pic  test.ipynb
/home/hmiwa/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
Basedata  StatGenSummerSchool2023_GenomeDataAnalysis5      targetdata
out.log   StatGenSummerSchool2023_GenomeDataAnalysis5.zip
/home/hmiwa/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials
Basedata  StatGenSummerSchool2023_GenomeDataAnalysis5      targetdata
out.log   StatGenSummerSchool2023_GenomeDataAnalysis5.zip


In [68]:
rawdir=~/hmiwa1/rawdata/Oceania
ls $rawdir

G05_sorted.genome.vcf.gz       T129_sorted.genome.vcf.gz.tbi
G05_sorted.genome.vcf.gz.tbi   T134_sorted.genome.vcf.gz
G06_sorted.genome.vcf.gz       T134_sorted.genome.vcf.gz.tbi
G06_sorted.genome.vcf.gz.tbi   t138_sorted.genome.vcf.gz
G135_sorted.genome.vcf.gz      t138_sorted.genome.vcf.gz.tbi
G135_sorted.genome.vcf.gz.tbi  T150_sorted.genome.vcf.gz
G150_sorted.genome.vcf.gz      T150_sorted.genome.vcf.gz.tbi
G150_sorted.genome.vcf.gz.tbi  T153_sorted.genome.vcf.gz
G152_sorted.genome.vcf.gz      T153_sorted.genome.vcf.gz.tbi
G152_sorted.genome.vcf.gz.tbi  T156_sorted.genome.vcf.gz
G164_sorted.genome.vcf.gz      T156_sorted.genome.vcf.gz.tbi
G164_sorted.genome.vcf.gz.tbi  t159_sorted.genome.vcf.gz
G217_sorted.genome.vcf.gz      t159_sorted.genome.vcf.gz.tbi
G217_sorted.genome.vcf.gz.tbi  T15_sorted.genome.vcf.gz
G220_sorted.genome.vcf.gz      T15_sorted.genome.vcf.gz.tbi
G220_sorted.genome.vcf.gz.tbi  t160_sorted.genome.vcf.gz
G256_sorted.genome.vcf.gz      t160_sorted.genome.vcf.gz.t

#### QCのチェックリスト

* Sample size
* File transfer
* Genome build
* Standard GWAS QC
* Ambiguous SNPs
* Mismatching SNPs
* Duplicate SNPs
* Sex Chromosomes
* Sample Overlap
* Relatedness
* Generate final QC'ed target data file

In [65]:
mkdir -p targetdata

##### Sample size

PRS分析は本来100人以上のターゲットデータが望ましいが、残念ながら20人くらいしかいない。致し方あるまい。

##### File transfer

md5sumコードに関しては今回は設定していない。次回以降準備しておいた方が良いかもね。

##### Genome build

これはhg19でタイピングされている。Basedataと一致している。

##### Standard GWAS QC

まずはトンガ人サンプルのQCを行う。

In [85]:
ls
ls $strdir

Basedata
StatGenSummerSchool2023_GenomeDataAnalysis5
StatGenSummerSchool2023_GenomeDataAnalysis5.zip
targetdata


Basedata
StatGenSummerSchool2023_GenomeDataAnalysis5
StatGenSummerSchool2023_GenomeDataAnalysis5.zip
targetdata


In [71]:
ls $rawdir/T*vcf.gz
ls $rawdir/t*vcf.gz

/home/hmiwa/hmiwa1/rawdata/Oceania/T03_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T104_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T105_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T106_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T109_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T120_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T129_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T134_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T150_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T153_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T156_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T15_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T26_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T33_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T47_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/T73_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oce

/home/hmiwa/hmiwa1/rawdata/Oceania/t112_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t138_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t159_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t160_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t21_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t38_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t51_sorted.genome.vcf.gz
/home/hmiwa/hmiwa1/rawdata/Oceania/t69_sorted.genome.vcf.gz


In [78]:
#準備
bcftools view --header-only $rawdir/T03_sorted.genome.vcf.gz | tail -n 1
bcftools view --no-header $rawdir/T03_sorted.genome.vcf.gz | head -n 1

#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	T03
chrM	1	.	G	.	.	LowGQX	.	GT:GQX:DP:DPF:MIN_DP	0/0:12:5:0:5


In [88]:
#準備（ツールの変数設定）
vcftools=/home/hmiwa/homebrew/bin/vcftools

データフォルダから読み出して、FILTER=PASSのものをGTだけにする

In [96]:
cd $strdir/targetdata
for t in T03 T104 T105 T106 T109 T120 T129 T134 T150 T153 T156 T15 T26 T33 T47 T73 T74 T76 T82 T85 T87 T88 T98 t112 t138 t159 t160 t21 t38 t51 t69
do
$vcftools --gzvcf $rawdir/${t}_sorted.genome.vcf.gz --remove-filtered-all --recode --stdout | bgzip -c > ${t}_sorted.genome_PASS_only.vcf.gz
bcftools index ${t}_sorted.genome_PASS_only.vcf.gz
for i in {1..22}
do
bcftools view --header-only  ${t}_sorted.genome_PASS_only.vcf.gz > ${t}_sorted.genome_PASS_only.chr${i}.vcf
bcftools query -r chr$i -f '%CHROM\t%POS\t%ID\t%REF\t%ALT\t%QUAL\t%FILTER\t%INFO\tGT[\t%GT]\n' ${t}_sorted.genome_PASS_only.vcf.gz | sed s/^chr//g >> ${t}_sorted.genome_PASS_only.chr${i}.vcf
bgzip -c ${t}_sorted.genome_PASS_only.chr${i}.vcf > ${t}_sorted.genome_PASS_only.chr${i}.vcf.gz
rm ${t}_sorted.genome_PASS_only.chr${i}.vcf
bcftools index ${t}_sorted.genome_PASS_only.chr${i}.vcf.gz
done
done

トンガ人サンプルをマージしたファイルを作る

In [97]:
cd $strdir/targetdata
for i in {1..22}
do
f1=`for t in T03 T104 T105 T106 T109 T120 T129 T134 T150 T153 T156 T15 T26 T33 T47 T73 T74 T76 T82 T85 T87 T88 T98 t112 t138 t159 t160 t21 t38 t51 t69; do echo "${t}_sorted.genome_PASS_only.chr${i}.vcf.gz "; done`
f=`echo $f1 | awk '{printf $0}'`
`echo "bcftools merge --missing-to-ref -r ${i} -Oz -o Tongan.chr${i}.vcf.gz ${f}"`
bcftools index Tongan.chr${i}.vcf.gz
done

[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types
[W::bcf_hdr_merge] Trying to combine "GQ" tag definitions of different types

マニュアルで推奨されているplinkコマンドは以下の通り

    plink \
        --bfile EUR \
        --maf 0.01 \
        --hwe 1e-6 \
        --geno 0.01 \
        --mind 0.01 \
        --write-snplist \
        --make-just-fam \
        --out EUR.QC

Beagleなどで処理する必要があるのでMAFやHWEなどはこの段階である程度フィルタリングしておく


Beagle処理の前にfixploidyをしておく必要がある

In [101]:
bcftools=/home/hmiwa/homebrew/bin/bcftools

In [102]:
cd $strdir/targetdata
for i in {1..22}
do
$vcftools --gzvcf Tongan.chr${i}.vcf.gz \
--hwe 0.000001  --max-missing 0.99 --maf 0.01 \
--remove-indels --min-alleles 2 --max-alleles 2 \
--remove-filtered-all --recode --stdout | bgzip -c > Tongan.chr${i}.QC.vcf.gz
$bcftools index Tongan.chr${i}.QC.vcf.gz
$bcftools +fixploidy Tongan.chr${i}.QC.vcf.gz -Oz --threads 24 -o Tongan.chr${i}.QC.FIXP.vcf.gz -- --default-ploidy 2
$bcftools index Tongan.chr${i}.QC.FIXP.vcf.gz
done

Beagle関係のディレクトリを変数に入れておく

In [103]:
beagle=/home/hmiwa/local/beagle.22Jul22.46e.jar
beagle_map=/home/hmiwa/local/plink.GRCh37.map

Beagleでフェージング処理をしておく

In [104]:
cd $strdir/targetdata
for i in {1..22}
do
java -Xmx16g -jar $beagle \
gt=Tongan.chr${i}.QC.FIXP.vcf.gz \
out=Tongan.chr${i}.QC.FIXP.phased \
map=${beagle_map}/plink.chr${i}.GRCh37.map \
nthreads=16 | tee Tongan.chr${i}.QC.FIXP.phased.log
done

beagle.22Jul22.46e.jar (version 5.4)
Copyright (C) 2014-2022 Brian L. Browning
Enter "java -jar beagle.22Jul22.46e.jar" to list command line argument
Start time: 10:55 PM JST on 18 Dec 2023

Command line: java -Xmx16384m -jar beagle.22Jul22.46e.jar
  gt=Tongan.chr1.QC.FIXP.vcf.gz
  out=Tongan.chr1.QC.FIXP.phased
  map=/home/hmiwa/local/plink.GRCh37.map/plink.chr1.GRCh37.map
  nthreads=16

Reference samples:                    0
Study     samples:                   31

Window 1 [1:10450-18870356]
Study     markers:               71,192

Burnin  iteration 1:           3 seconds
Burnin  iteration 2:           1 second
Burnin  iteration 3:           1 second

Estimated ne:                  24080
Estimated err:                 8.9e-03

Phasing iteration 1:           1 second
Phasing iteration 2:           0 seconds
Phasing iteration 3:           0 seconds
Phasing iteration 4:           0 seconds
Phasing iteration 5:           0 seconds
Phasing iteration 6:           0 seconds
Phasing iterat

そのまえにこの段階でSNP IDを付与しておかねばならない。まず前段階の準備を行う。

In [105]:
cd $strdir/targetdata
#SNPID付与の準備（中身はVCF形式なので拡張子の変更をする）
gunzip -c ~/hmiwa1/rawdata/SNP/GCF_000001405.25.gz | bgzip -c > GCF_000001405.25.vcf.gz
bcftools index GCF_000001405.25.vcf.gz

In [106]:
#上記はCHROMがalias sequence namesになっているので変更する
wget http://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/hg19.chromAlias.txt
wget https://raw.githubusercontent.com/humgendiv/hmiwa-shared/main/get_rsID/aliasrename.R
Rscript aliasrename.R

echo "start annotating"
bcftools annotate --rename-chrs alias.new.ids \
-Oz -o GCF_000001405.25.annotated.vcf.gz GCF_000001405.25.vcf.gz
echo "... annotated"
echo "start indexing"
bcftools index GCF_000001405.25.annotated.vcf.gz
echo "... indexed"

--2023-12-19 00:57:13--  http://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/hg19.chromAlias.txt
Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163
Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 16866 (16K) [text/plain]
Saving to: ‘hg19.chromAlias.txt’

hg19.chromAlias.txt 100%[===================>]  16.47K  --.-KB/s    in 0.1s    

2023-12-19 00:57:14 (154 KB/s) - ‘hg19.chromAlias.txt’ saved [16866/16866]

--2023-12-19 00:57:14--  https://raw.githubusercontent.com/humgendiv/hmiwa-shared/main/get_rsID/aliasrename.R
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 402 [text/plain]
Saving to: ‘aliasrename.R’

aliasrename.R    

In [107]:
#染色体ごとに分割する
echo "start separating by chromosomes"
for i in {1..22}
do
bcftools view -r chr${i} GCF_000001405.25.annotated.vcf.gz | \
sed -e s/^chr//g | bgzip -c > GCF_000001405.25.annotated.chr${i}.vcf.gz
echo "... chr${i} separated"
bcftools index GCF_000001405.25.annotated.chr${i}.vcf.gz
echo "... chr${i} indexed"
done

start separating by chromosomes
... chr1 separated
... chr1 indexed
... chr2 separated
... chr2 indexed
... chr3 separated
... chr3 indexed
... chr4 separated
... chr4 indexed
... chr5 separated
... chr5 indexed
... chr6 separated
... chr6 indexed
... chr7 separated
... chr7 indexed
... chr8 separated
... chr8 indexed
... chr9 separated
... chr9 indexed
... chr10 separated
... chr10 indexed
... chr11 separated
... chr11 indexed
... chr12 separated
... chr12 indexed
... chr13 separated
... chr13 indexed
... chr14 separated
... chr14 indexed
... chr15 separated
... chr15 indexed
... chr16 separated
... chr16 indexed
... chr17 separated
... chr17 indexed
... chr18 separated
... chr18 indexed
... chr19 separated
... chr19 indexed
... chr20 separated
... chr20 indexed
... chr21 separated
... chr21 indexed
... chr22 separated
... chr22 indexed


アノテーションを実行する。

In [1]:
ls

05_231211_RePRSTrials.ipynb  Documents  jupyter_venv5  pic  test.ipynb


In [5]:
cd $strdir/targetdata

echo START: `date`
#常染色体すべてにrsIDを付与する場合
for i in {1..22}
do
INFILE=Tongan.chr${i}.QC.FIXP.phased.vcf.gz
REFFILE=GCF_000001405.25.annotated.chr${i}.vcf.gz
OUTFILE=Tongan.chr${i}.QC.FIXP.phased.annotated.vcf.gz
bcftools index $INFILE
bcftools annotate -a $REFFILE -Oz -o $OUTFILE -c ID $INFILE --threads 16
bcftools index $OUTFILE
done
echo END: `date`

START: 2023年 12月 19日 火曜日 17:08:39 JST
[E::main_vcfindex] the index file exists. Please use '-f' to overwrite Tongan.chr22.QC.FIXP.phased.vcf.gz.csi
END: 2023年 12月 19日 火曜日 17:25:20 JST


常染色体をひとつのファイルにまとめる

In [7]:
bcftools concat \
    -Oz -o Tongan.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr1.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr2.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr3.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr4.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr5.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr6.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr7.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr8.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr9.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr10.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr11.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr12.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr13.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr14.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr15.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr16.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr17.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr18.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr19.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr20.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr21.QC.FIXP.phased.annotated.vcf.gz \
    Tongan.chr22.QC.FIXP.phased.annotated.vcf.gz
bcftools index Tongan.QC.FIXP.phased.annotated.vcf.gz

Checking the headers and starting positions of 22 files
Concatenating Tongan.chr1.QC.FIXP.phased.annotated.vcf.gz	2.630552 seconds
Concatenating Tongan.chr2.QC.FIXP.phased.annotated.vcf.gz	2.533487 seconds
Concatenating Tongan.chr3.QC.FIXP.phased.annotated.vcf.gz	2.074929 seconds
Concatenating Tongan.chr4.QC.FIXP.phased.annotated.vcf.gz	2.101666 seconds
Concatenating Tongan.chr5.QC.FIXP.phased.annotated.vcf.gz	1.768070 seconds
Concatenating Tongan.chr6.QC.FIXP.phased.annotated.vcf.gz	1.859544 seconds
Concatenating Tongan.chr7.QC.FIXP.phased.annotated.vcf.gz	1.723798 seconds
Concatenating Tongan.chr8.QC.FIXP.phased.annotated.vcf.gz	1.509193 seconds
Concatenating Tongan.chr9.QC.FIXP.phased.annotated.vcf.gz	1.394220 seconds
Concatenating Tongan.chr10.QC.FIXP.phased.annotated.vcf.gz	1.418890 seconds
Concatenating Tongan.chr11.QC.FIXP.phased.annotated.vcf.gz	1.395065 seconds
Concatenating Tongan.chr12.QC.FIXP.phased.annotated.vcf.gz	1.363414 seconds
Concatenating Tongan.chr13.QC.FIXP.phased

ここからはPLINK形式のファイルで作業したいので変換しておく。

In [8]:
plink \
    --vcf Tongan.QC.FIXP.phased.annotated.vcf.gz \
    --make-bed \
    --out TNG

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.log.
Options in effect:
  --make-bed
  --out TNG
  --vcf Tongan.QC.FIXP.phased.annotated.vcf.gz

257564 MB RAM detected; reserving 128782 MB for main workspace.
--vcf: TNG-temporary.bed + TNG-temporary.bim + TNG-temporary.fam written.
8711185 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.nosex .
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
8711185 variants and 31 people pass filters and QC.
Note: No phenoty

In [9]:
ls TNG*

TNG.bed  TNG.bim  TNG.fam  TNG.log  TNG.nosex


ジェノタイピング率が低いSNP、マイナーアレル頻度が低いSNP、ハーディーワインベルグ平衡から外れたSNP、ジェノタイピング率が低い個体などの除去を行う。

In [10]:
plink \
    --bfile TNG \
    --maf 0.01 \
    --hwe 1e-6 \
    --geno 0.01 \
    --mind 0.01 \
    --write-snplist \
    --make-just-fam \
    --out TNG.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.QC.log.
Options in effect:
  --bfile TNG
  --geno 0.01
  --hwe 1e-6
  --maf 0.01
  --make-just-fam
  --mind 0.01
  --out TNG.QC
  --write-snplist

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.QC.nosex .
0 people removed due to missing genotype data (--mind).
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
0 variants removed due to missing

In [11]:
ls TNG*

TNG.bed  TNG.fam  TNG.nosex   TNG.QC.log    TNG.QC.snplist
TNG.bim  TNG.log  TNG.QC.fam  TNG.QC.nosex


極端なヘテロ接合度を持つ個体を除去する。

まずは相関性の高いSNPを除去するためにプルーニング

In [12]:
plink \
    --bfile TNG \
    --keep TNG.QC.fam \
    --extract TNG.QC.snplist \
    --indep-pairwise 200 50 0.25 \
    --out TNG.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.QC.log.
Options in effect:
  --bfile TNG
  --extract TNG.QC.snplist
  --indep-pairwise 200 50 0.25
  --keep TNG.QC.fam
  --out TNG.QC

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.QC.nosex .
--extract: 8708432 variants remaining.
--keep: 31 people remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
8708432 variants and 31 people pa

ヘテロ接合度を計算する

In [13]:
plink \
    --bfile TNG \
    --extract TNG.QC.prune.in \
    --keep TNG.QC.fam \
    --het \
    --out TNG.QC

PLINK v1.90b7 64-bit (16 Jan 2023)             www.cog-genomics.org/plink/1.9/
(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to TNG.QC.log.
Options in effect:
  --bfile TNG
  --extract TNG.QC.prune.in
  --het
  --keep TNG.QC.fam
  --out TNG.QC

257564 MB RAM detected; reserving 128782 MB for main workspace.
8711185 variants loaded from .bim file.
31 people (0 males, 0 females, 31 ambiguous) loaded from .fam.
Ambiguous sex IDs written to TNG.QC.nosex .
--extract: 660791 variants remaining.
--keep: 31 people remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 31 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is exactly 1.
660791 variants and 31 people pass filters and QC.
Note:

現在のディレクトリチェック

In [4]:
cd $strdir
ls

Basedata  StatGenSummerSchool2023_GenomeDataAnalysis5      targetdata
out.log   StatGenSummerSchool2023_GenomeDataAnalysis5.zip


ところでVScodeでjupyterのカーネルが選べないときはjupyterの拡張機能のバージョンを落としてみてくださいね（蛇足）

In [5]:
cd $strdir
cat << EOL > $gitdir/het.R
setwd("$strdir/targetdata")
dat <- read.table("TNG.QC.het", header=T) # Read in the EUR.het file, specify it has header
m <- mean(dat\$F) # Calculate the mean
s <- sd(dat\$F) # Calculate the SD
valid <- subset(dat, F <= m+3*s & F >= m-3*s) # Get any samples with F coefficient within 3 SD of the population mean
write.table(valid[,c(1,2)], "TNG.valid.sample", quote=F, row.names=F) # print FID and IID for valid samples
EOL
cat $gitdir/het.R

setwd("/home/hmiwa/hmiwa1/storage/hmiwa-lab/05_231211_RePRSTrials/targetdata")
dat <- read.table("TNG.QC.het", header=T) # Read in the EUR.het file, specify it has header
m <- mean(dat$F) # Calculate the mean
s <- sd(dat$F) # Calculate the SD
valid <- subset(dat, F <= m+3*s & F >= m-3*s) # Get any samples with F coefficient within 3 SD of the population mean
write.table(valid[,c(1,2)], "TNG.valid.sample", quote=F, row.names=F) # print FID and IID for valid samples


In [6]:
Rscript --no-save $gitdir/het.R

In [8]:
cd $strdir/targetdata
ls TNG*

TNG.bed  TNG.log     TNG.QC.het    TNG.QC.prune.in   TNG.valid.sample
TNG.bim  TNG.nosex   TNG.QC.log    TNG.QC.prune.out
TNG.fam  TNG.QC.fam  TNG.QC.nosex  TNG.QC.snplist


In [9]:
head TNG.valid.sample

FID IID
T03 T03
T104 T104
T105 T105
T106 T106
T109 T109
T120 T120
T129 T129
T134 T134
T150 T150


In [10]:
wc TNG.valid.sample
wc TNG.QC.het

 32  64 284 TNG.valid.sample
  32  192 1984 TNG.QC.het


##### Ambiguous SNPs

ここはBaseDataで削除済みなのでOK

##### Mismatching SNPs

##### Duplicate SNPs

##### Sex Chromosomes

##### Sample Overlap

##### Relatedness

##### Generate final QC'ed target data file